### ISSUE TITLE
CreateContextAttribs always fails on windows

### ISSUE THREAD

**Issue by Ollhax** [opentk/opentk#754]
### Description

Creating the context via CreateContextAttribs always fails on windows, which triggers it to fall back to CreateContext. I've tracked down the problem to WinGLContext.cs, where it adds a profile mask attribute:

```
if (flags != 0)
{
	attributes.Add((int)ArbCreateContext.ContextFlags);
	attributes.Add((int)GetARBContextFlags(flags));
	attributes.Add((int)ArbCreateContext.ProfileMask);
	attributes.Add((int)GetARBContextProfile(flags)); // <-- PROBLEM HERE
}

private static ArbCreateContext GetARBContextFlags(GraphicsContextFlags flags)
{
	ArbCreateContext result = 0;
	result |= (flags & GraphicsContextFlags.ForwardCompatible) != 0 ?
		ArbCreateContext.CoreProfileBit : ArbCreateContext.CompatibilityProfileBit;
	return result;
}

private static ArbCreateContext GetARBContextProfile(GraphicsContextFlags flags)
{
	ArbCreateContext result = 0;
	result |= (flags & GraphicsContextFlags.Debug) != 0 ? ArbCreateContext.DebugBit : 0;
	return result;
}
```

If I'm reading the [specs](https://www.khronos.org/registry/OpenGL/extensions/ARB/WGL_ARB_create_context.txt) right: "If attribute WGL_CONTEXT_PROFILE_MASK_ARB has no bits set; (..) then ERROR_INVALID_PROFILE_ARB is generated." - i.e. GetARBContextProfile is incorrect, since it returns a zero when not passing the debug flag.

In addition, I think GetARBContextFlags is confused, returning profile bit settings rather than either of the two valid settings, WGL_CONTEXT_DEBUG_BIT_ARB and WGL_CONTEXT_FORWARD_COMPATIBLE_BIT_ARB.

I'm not sure exactly how OpenTK should interpret the user settings, but I think the two methods should look somewhat like this:

```
static ArbCreateContext GetARBContextFlags(GraphicsContextFlags flags)
{
	ArbCreateContext result = 0;
	result |= (flags & GraphicsContextFlags.ForwardCompatible) != 0 ? ArbCreateContext.ForwardCompatibleBit : 0;
	result |= (flags & GraphicsContextFlags.Debug) != 0 ? ArbCreateContext.DebugBit : 0;
	return result;
}

static ArbCreateContext GetARBContextProfile(GraphicsContextFlags flags)
{
	ArbCreateContext result = 0;
	// Interpret the ForwardCompatible flag as wanting the core profile. (Should make a separate flag to differentiate core profile vs forward compatible one?)
	result |= (flags & GraphicsContextFlags.ForwardCompatible) != 0 ? ArbCreateContext.CoreProfileBit : ArbCreateContext.CompatibilityProfileBit;
	return result;
}
```

### Related information 

* Windows 10
* Native backend
* A slightly customized version of OpenTK 2.0.0, built from source
* .NET Framework

**Comment by varon** [comment]
This seems kinda weird.

Would need a fully complete code example to check. Due to the age, I'll presume you've given up/solved it, and will close.

Re-open it if you want more help.

**Comment by Ollhax** [comment]
Well, the code example is in WinGLContext.cs, at least up to OpenTK 3.0. Just try to create e.g. a GameWindow (make sure to not specify GraphicsContextFlags.Debug) and note that the context creation fails and uses a fallback instead. If I'm right about this, it's just set up wrong. It incorrectly sends the Debug context bit to the Profile attribute, and Profile context bit to the Flags attribute.

I don't really need any help, I have a build of OpenTK where I've implemented this fix. I just wanted to share the solution. I would make it a pull request, but as I noted in my comments I'm not sure how to interpret the GraphicsContextFlags. With only the flag GraphicsContextFlags.ForwardCompatible that sets both the Core and ForwardCompatibility bits, there's no way to create a context that is Core but not ForwardCompatible. I don't really think this is an issue, but it needs some consideration. E.g:

- From what I read a Core context without the ForwardCompatibility bit set is almost the same as having it set, with a few exceptions. E.g. a Core, non-ForwardCompatible context can use "wide lines", which is deprecated but not removed in the Core context.
- [Apparently](https://www.reddit.com/r/gamedev/comments/27c93z/os_x_opengl_corecompatibility/) the context creation crashes on OSX if you try to create a Core, non-ForwardCompatible context.
- Adding GraphicsContextFlags.Core would break a lot of code that uses OpenTK. It would be an easy fix, but still.

IMO it's probably for the best to just interpret GraphicsContextFlags.ForwardCompatible as wanting a Core profile (like in my example above), but its your call.

**Comment by Ollhax** [comment]
@varon Just pinging in case you didn't see.

**Comment by DylanFPS** [comment]
<img src="https://cdn.discordapp.com/emojis/437088256291504130.png?v=1" width="50" />

**Comment by varon** [comment]
Thanks, @Ollhax - It's really great to see a reply on this. We should be able to address this for 4.0.

**Comment by varon** [comment]
Should be fixed by #935
